streamer/periodconcat_node.py

```diff
@@ -0,0 +1,318 @@
+    # We will start the concatenation once all the PackagerNode(s) are finished.
+    for node in all_nodes:
+      if isinstance(node, TranscoderNode):
+        self._transcoder_nodes.append(node)
+        self._period_inputs_list.append(node._inputs)
+      elif isinstance(node, PackagerNode):
+        self._packager_nodes.append(node)
+  
+  def _thread_single_pass(self) -> None:
+    """Watches all the PackagerNode(s), if at least one of them is running it skips this
+    _thread_single_pass, if all of them are finished, it starts period concatenation, if one of
+    them is errored, it raises a RuntimeError.
+    """
+    
+    for i, packager_node in enumerate(self._packager_nodes):
+      status: ProcessStatus = packager_node.check_status()
+      if status == ProcessStatus.RUNNING:
+        return
+      elif status == ProcessStatus.ERRORED:
+        raise RuntimeError(
+          "PeriodConcatNode: 'PackagerNode#{} Errored, Concatenation is stopped.'".format(i)
+        )
+    
+    if(ManifestFormat.DASH in self._pipeline_config.manifest_format):
+      self._dash_concat()
+      
+    if(ManifestFormat.HLS in self._pipeline_config.manifest_format):
+      self._hls_concat()
+    
+    self._status = ProcessStatus.FINISHED
+  
+  def _dash_concat(self) -> None:
+    """Concatenates multiple single-period DASH manifests into one multi-period DASH manifest."""
+    
+    class ISO8601:
+      """A helper class represeting the iso8601:2004 duration format."""
+      
+      def __init__(self, Y=0, PM=0, W=0, D=0, H=0, TM=0, S=0,
+                   duration: Optional[str] =None):
+        """If given a duration, it uses it to set the time containers."""
+        
+        self.PY: int = Y # Years
+        self.PM: int = PM  # Months
+        self.PW: int = W # Weeks
+        self.PD: int = D # Days
+        self.TH: int = H # Hours
+        self.TM: int = TM  # Minutes
+        self.TS: float = S # Seconds
+        self.dur = 'PT0S' # Duration string in iso8601
+        if duration:
+          self.parse(duration)
+        
+        # Update self.dur
+        self.update_dur()
+        
+      def update_dur(self) -> None:
+        """Write duration in iso8601 formate."""
+        
+        # (P) The duration designator
+        self.dur = 'P'
+        if self.PY:
+          self.dur += str(self.PY) + 'Y'
+        if self.PM:
+          self.dur += str(self.PM) + 'M'
+        if self.PW:
+          self.dur += str(self.PW) + 'W'
+        if self.PD:
+          self.dur += str(self.PD) + 'D'
+          
+        # (T) The time designator
+        self.dur += 'T'
+        if self.TH:
+          self.dur += str(self.TH) + 'H'
+        if self.TM:
+          self.dur += str(self.TM) + 'M'
+        if self.TS:
+          self.dur += str(self.TS) + 'S'
+        
+        # If all the time containers were zero, write zero seconds
+        if self.dur == 'PT':
+          self.dur += '0S'
```

Comment:
First, you can have 0 in these fields, so you can just have every value.  Second, since we target Python3, you can use format strings for this.

```py
return f'P{self.PY}Y{self.PM}M{self.PW}W{self.PD}DT{self.TH}H{self.TM}M{self.TS}S'
```

Reply:
yes i just wanted the output to look cleaner :).
i looked up for f strings and they where supported for >= python3.6, is that ok?